keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1348 @@
+Optimizations like caching or lazy sub-schema resolution can be candidates of beta/GA graduation criteria.
+
+#### Unsetting values
+
+Since there is no field manager used for the merge, the server side apply merge
+algorithm will only add and replace values. This is because server side apply is
+designed to only unset values that were previously owned by the field manager
+but excluded from the apply configuration.
+
+To work around this limitation, we will take advantage of CEL's optional type
+feature to make it possible to express that a value should be unset. For example:
+
+```cel
+Object{
+  spec: Object.spec{
+    ?fieldToRemoveIfPresent: optional.none()
+  }
+}
+```
+The policy declaratively describes that, in the final object, if fieldToRemoveIfPresent presents it should be removed or no-op otherwise.
+
+The optional.none() function creates a CEL object defined as optional(T) where T is the type of the receiving field. 
+In conjunction with standard CEL macros, simple expressions can perform more complicated and precise operations.
+
+For example, to remove the env of a sidecar container, filter by its name.
+
+```yaml
+mutations:
+  - mutaton: >
+        Object{
+            spec: Object.spec{
+                containers: object.spec.containers{
+                    object.spec.containers.filter(c, c.name == "sidecar")
+                    .map(c, Object.spec.containers.item{
+                        ?env: optional.none()
+                    })
+                }
+            }
+        }
+
+```
+
+We will track which fields are unset in this way and remove them after the
+server side apply merge algorithm is run.
+
+This solves the vast majority of value removal needs. Specifically:
+
+| Schema type | Merge type | Example of how to unset a value                                          |
+|-------------| -----------|--------------------------------------------------------------------------|
+| struct      | atomic     | ```Object{ spec: Object.spec{ structField: {?fieldToRemove: optional.none()}}}``` |
+| struct      | granular   | `?fieldToRemove: optional.none()`                                        |
+| map         | atomic     | `mapField: object.spec.mapField.filter(k, k != "keyToRemove")`           |
+| map         | granular   | `mapField: {?"keyToRemove": optional.none()}`                            |
+| list        | atomic     | `listField: object.spec.listField.filter(e, e != "itemToRemove")`        |
+| list        | set        | `setField: object.spec.setField.filter(e, e != "itemToRemove")`          |
+| list        | map        | See below                                                                |
+| list        | granular   | Not Supported                                                            | 
+<<[UNRESOLVED @cici37 ]>>
+If it is ok to not support granular list deletion <<[/UNRESOLVED]>>
```

Comment:
oh, is this my example from above: how do I remove a container with field X or named X?

Reply:
I think we have an invalid row in the table.  `x-kubernetes-list-type` can be set to `atomic` (default), `map` or `set`. We already have rows for these.  The `granular` entry here might be leading to confusion?

For "how do I remove an entire container with an envvar named remove-this-container?", containers (and initContainers) use `x-kubernetes-list-type=map`, and so we need a agreed solution for that case.

Reply:
For the use case on list map, we have suggested a way as listed in the kep. Another question is about deletion for regular list(e.g.: how to define the expression on deleting a couple args from a list of string args). User might be able to bypass it by deleting the entire list and add what they need in a separate mutation. Is that sufficient enough?

Reply:
> Another question is about deletion for regular list

`atomicList: object.spec.atomicList.filter()` should be sufficient for preserving everything in an atomic list except the items that one wants to delete.

Reply:
In the example of having a list [1,2,3] which we wanna delete 1 from the list, by using filter might construct a apply configuration patch like [2,3], by applying structured merge, it will result in [1,2,3] with no deletion applied. Is this understanding right? Or if you were talking about by treating it as atomic underlying we do a delete the entire atomicList and then apply the filter()? Is that ok to assume the list is atomic?

Reply:
An apply of [2, 3] to a list of [1, 2, 3] is designed to replace the list atomically, resulting in [2, 3]